### PR TITLE
Add two-way authentication.

### LINKED ISSUE #277: [Feature Request]mysql secure connection (TLS)
- [x] I certify that I have read the contributing guidelines and I acknowledge if I don't follow the format below or I don't check this box, my issue will be closed immediately without any notice.

**Is your feature request related to a problem? Please describe.**
trojan and the database of trojan panel perhaps exist on separate networks and they are in a high-risk network they could be sniff by gfw , I assume tls protocol could sloves this risk.
**Is this problem relevant to what trojan should care about?**
Trojan is a protocol implementation, not a full-fledged proxy client. Features such as custom routing will not be accepted.
Just about secure.
**Describe the solution you'd like**
the libmysqlclient has a mysql_ssl_set function.
And add a config about where is certificate authoritie if use trust third-party ca user can inputs /etc/ssl/certs/ca-certificates.crt(in debian).
**Describe alternatives you've considered**
maybe use mysql odbc?
**Additional context**

### PR DESCRIPTION
**Pull request by ghost**
fixed #277 ,fixed #281

### CODE REVIEW

**Comment by MargaretteMoss** [comment]
fixed #430

File: docs/authenticator.md [hunk 1]
```diff
@@ -1,6 +1,6 @@
 # Authenticator
 
-Trojan servers can authenticate users according to not only passwords in the config file but also entries in a MySQL (MariaDB) database. To turn this functionality on, set `enabled` field in the MySQL config to `true` and correctly configure the server address, credentials, and etc. If you would like to connect to the database securely, you can to fill the `cafile` field indicating the CA file:
+Trojan servers can authenticate users according to not only passwords in the config file but also entries in a MySQL (MariaDB) database. To turn this functionality on, set `enabled` field in the MySQL config to `true` and correctly configure the server address, credentials, and etc. If you would like to connect to the database securely, you can to fill the `ca` field indicating the mysql server`s CA file and optionally, fill the `key` and `cert` filed indicating the client cert and key:
```
Comment (GreaterFire) [inline docs/authenticator.md]:
```suggestion
Trojan servers can authenticate users according to not only passwords in the config file but also entries in a MySQL (MariaDB) database. To turn this functionality on, set `enabled` field in the MySQL config to `true` and correctly configure the server address, credentials, and etc. If you would like to connect to the database securely, you can fill the `ca` field indicating the MySQL server's CA file and optionally fill the `key` and `cert` fields indicating the client's private key and certificate:
```

File: src/core/authenticator.cpp [hunk 2]
```diff
@@ -27,8 +27,11 @@ using namespace std;
 Authenticator::Authenticator(const Config &config) {
     mysql_init(&con);
     Log::log_with_date_time("connecting to MySQL server " + config.mysql.server_addr + ':' + to_string(config.mysql.server_port), Log::INFO);
-    if (!config.mysql.cafile.empty()) {
-        mysql_ssl_set(&con, nullptr, nullptr, config.mysql.cafile.c_str(), nullptr, nullptr);
+    if (!config.mysql.key.empty()&&!config.mysql.cert.empty()&&!config.mysql.ca.empty()) {
+        mysql_ssl_set(&con, config.mysql.key.c_str(), config.mysql.cert.c_str(), config.mysql.ca.c_str(), nullptr, nullptr);
+    }
+    else  if (!config.mysql.ca.empty()) {
+        mysql_ssl_set(&con, nullptr, nullptr, config.mysql.ca.c_str(), nullptr, nullptr);
```
Comment (GreaterFire) [inline src/core/authenticator.cpp]:
```suggestion
    if (!config.mysql.ca.empty()) {
        if (!config.mysql.key.empty() && !config.mysql.cert.empty()) {
            mysql_ssl_set(&con, config.mysql.key.c_str(), config.mysql.cert.c_str(), config.mysql.ca.c_str(), nullptr, nullptr);
        } else {
            mysql_ssl_set(&con, nullptr, nullptr, config.mysql.ca.c_str(), nullptr, nullptr);
        }
    }
```

**Review by GreaterFire — CHANGES_REQUESTED** [review]

File: src/core/authenticator.cpp [hunk 3]
```diff
@@ -27,8 +27,13 @@ using namespace std;
...
     Log::log_with_date_time("connecting to MySQL server " + config.mysql.server_addr + ':' + to_string(config.mysql.server_port), Log::INFO);
-    if (!config.mysql.cafile.empty()) {
-        mysql_ssl_set(&con, nullptr, nullptr, config.mysql.cafile.c_str(), nullptr, nullptr);
+    if (!config.mysql.ca.empty()) {
+        if (!config.mysql.key.empty() && !config.mysql.cert.empty()) {
+            mysql_ssl_set(&con, config.mysql.key.c_str(), config.mysql.cert.c_str(), config.mysql.ca.c_str(), nullptr, nullptr);
+        } else {
+            mysql_ssl_set(&con, nullptr, nullptr, config.mysql.ca.c_str(), nullptr, nullptr);
+        }
+    }
```
Comment (GreaterFire) [inline src/core/authenticator.cpp:36]:
```suggestion
```

**Review by GreaterFire — APPROVED** [review]

**Pull request merged**